compiler/qsc_linter/src/lints/hir.rs

```diff
@@ -1,17 +1,80 @@
-use qsc_hir::hir::Lit;
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// use miette::Diagnostic;
+use qsc_data_structures::span::Span;
+use qsc_hir::{
+    hir::{CallableDecl, CallableKind, Expr, ExprKind, Stmt, StmtKind},
+    ty::Ty,
+    visit::{self, Visitor},
+};
+// use thiserror::Error;
 
 use crate::linter::hir::declare_hir_lints;
 
 use super::lint;
 
 declare_hir_lints! {
-    (Placeholder, LintLevel::Allow, "this a placeholder", "remove after addding the first HIR lint"),
+    (NeedlessOperation, LintLevel::Warn, "unnecessary operation declaration", "convert to function")
+}
+
+#[derive(Default)]
+pub(super) struct OperationLimits {
+    // Operation Characteristics
+    pub(super) op_char: Vec<Span>,
+}
+
+impl Visitor<'_> for OperationLimits {
+    fn visit_callable_decl(&mut self, decl: &CallableDecl) {
+        if decl.kind == CallableKind::Operation {
+            // Checking for special Ctl and Adj implementations
+            // if decl.adj.is_some() || decl.ctl.is_some() || decl.ctl_adj.is_some() {
+            //     self.op_char.push(decl.span);
+            // }
+            // if decl.functors != FunctorSetValue::Empty {
+            //     self.op_char.push(decl.name.span);
+            // }
```

Comment:
I think it's OK not to show the lint if you have an `operation` that defines explicit specializations. 

Most importantly, this is so that our own code in the standard library compiles without warnings:
https://github.com/microsoft/qsharp/blob/9aa8aaff64e08d481171b71245d5a072350f0adf/library/std/intrinsic.qs#L184-L193

But we can show the lint if you have _implicit_ specializations. Funny enough, this means that our own specialization sample would need to be fixed.
https://github.com/microsoft/qsharp/blob/9aa8aaff64e08d481171b71245d5a072350f0adf/samples/language/Specializations.qs#L12-L16

A user who wants to deliberately override the lint can define an explicit specialization. Maybe that can be mentioned in the error help text.

Reply:
Thanks for thinking of these edge cases, you're uncovering some stuff that hadn't occurred to me before!

After some discussion within the team, we realized that "empty operations" could also be a special case that we exclude from the lint. That takes care of `I` and `DoNothing`.

So the overall set of rules would be (unless I'm missing something)

- empty operations: no lint
- operations with errors (e.g. partially typed code): no lint*
- non-empty operations, with specializations, and no quantum operations: show lint, but don't offer quickfix (to avoid deleting user code in any explicit specializations)
- non-empty operations with _no_ specializations, and no quantum operations: show lint, offer quickfix to convert to function

*I don't think the linter runs at all if there are errors, so I don't think this needs special consideration.

Reply:
Sorry, one more to check! Lambdas can be declared as `operation`s or `function`s, differentiated by the arrow syntax.


```qsharp
let a = (q) => CNOT(q); // operation
let a = (q) -> CNOT(q); // function (error)
let a = (a) => a + 1;   // declared as operation, but could be function
let a = (a) -> a + 1; 
```

Ideally the 3rd one would be a lint. (Even if we can't support these fully, at the very least, we want to verify the behavior is reasonable.)

[Playground](https://microsoft.github.io/qsharp/?code=H4sIAAAAAAAACm2NQQ6CMBRE9z3FLNsoIW4lGBPjEsVED%2FAp1ZBAC6VdEMLdDV1AjPzln5n3NDWqb0kqZMPDk3a%2Bya35WGowMgYA56t2dshNpR0X4WNaZclVRiOjSnOBI166chhDOl%2BtHAgpeCeQnnC53Z%2B8EwnieB1vlqPf8ttrGURcWWus%2BN9QEBB2OCTAvCmVrMmqEtSvsj0K7yCNr0sUauFu8qKFF%2BKJTewLZ4XROCYBAAA%3D&profile=unrestricted)

Reply:
I'd consider quick fix out of scope for the purposes of the bounty, but it certainly would be nice to have to make the feature feel complete!